lib/Conversion/StandardToHandshake/MaximizeSSA.cpp

```diff
@@ -0,0 +1,230 @@
+//
+// Contains the definitions of the SSA maximization pass as well as utilities
+// for converting a function with standard control flow into maximal SSA form.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../PassDetail.h"
+#include "circt/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <stack>
+
+using namespace mlir;
+using namespace circt;
+
+static Block *getDefiningBlock(Value value) {
+  // Value is either a block argument...
+  if (auto blockArg = dyn_cast<BlockArgument>(value); blockArg)
+    return blockArg.getParentBlock();
+
+  // ... or an operation's result
+  auto *defOp = value.getDefiningOp();
+  assert(defOp);
+  return defOp->getBlock();
+}
+
+static LogicalResult addArgToTerminator(Block *block, Block *predBlock,
+                                        Value value) {
+
+  // Identify terminator branching instruction in predecessor block
+  auto branchOp = dyn_cast<BranchOpInterface>(predBlock->getTerminator());
+  if (!branchOp)
+    return predBlock->getTerminator()->emitError(
+        "Expected terminator operation of block to be a "
+        "branch-like operation");
+
+  // In the predecessor block's terminator, find all successors that equal
+  // the block and add the value to the list of operands it's passed
+  auto successors = branchOp->getSuccessors();
+  for (uint idx = 0, numSucc = successors.size(); idx < numSucc; ++idx) {
+    auto *succBlock = successors[idx];
+    if (succBlock == block) {
+      branchOp.getSuccessorOperands(idx).append(value);
+    }
+  }
+
+  return success();
+}
+
+namespace circt {
+
+bool isRegionSSAMaximized(Region &region) {
+
+  // Check whether all operands used within each block are also defined within
+  // the same block
+  for (auto &block : region.getBlocks())
+    for (auto &op : block.getOperations())
+      for (auto operand : op.getOperands())
+        if (getDefiningBlock(operand) != &block)
+          return false;
+
+  return true;
+}
+
+LogicalResult maximizeSSA(Value value, PatternRewriter &rewriter) {
+
+  // Identify the basic block in which the value is defined
+  Block *defBlock = getDefiningBlock(value);
+
+  // Identify all basic blocks in which the value is used (excluding the
+  // value-defining block)
+  DenseSet<Block *> blocksUsing;
+  for (auto &use : value.getUses()) {
+    auto *block = use.getOwner()->getBlock();
+    if (block != defBlock)
+      blocksUsing.insert(block);
+  }
+
+  // Prepare a stack to store the list of basic blocks that must be modified for
+  // the value to be in maximum SSA form. At all points, blocksUsing is a
+  // non-strict superset of the elements contained in blocksToVisit
+  std::stack<Block *> blocksToVisit;
+  for (auto *block : blocksUsing) {
```

Comment:
Cleaner indeed, thanks!

Reply:
`llvm::copy` expects something like a `std::back_inserter` or an `std::inserter`. Not sure if the latter works with stacks. 
As `stack` should not give any reasonable advantage over `SmallVector`, this might be a good reason to switch.